aten/src/ATen/core/TensorMethods.h

```diff
@@ -1240,6 +1237,21 @@ inline Device Tensor::device() const {
   return Device(type().device_type(), type().is_cuda() ? get_device() : -1);
 }
 
+inline int64_t Tensor::get_device() const {
+  // NB: This gets called a lot so we special case it here instead of dispatching
+  // to a native function.
+  const auto& mytype = type();
+  if (!mytype.is_cuda()) {
+    return -1;
+  }
+  if (mytype.is_sparse()) {
+    return _values().get_device();
+  }
+  // TODO(rzou): Investigate caching device on TensorImpl.
+  // It'll probably be faster than having this indirection through Storage
+  return impl_->storage().unsafeGetStorageImpl()->device().index();
```

Comment:
Err... I'm pretty sure you don't need `unsafeGetStorageImpl()` here; we should give you `device()` directly on Storage.